src/qibo/derivative.py

```diff
@@ -102,6 +103,12 @@ def circuit(nqubits = 1):
 
     # inheriting hamiltonian's backend
     backend = hamiltonian.backend
+    # TODO: make this work wih pytorch backend
+    if isinstance(backend, PyTorchBackend):
+        raise_error(
+            NotImplementedError,
+            "PyTorchBackend for the parameter shift rule is not supported.",
+        )
```

Comment:
This #TODO is probably one of the most important things here, so it should be done before merging

Reply:
@Simone-Bordoni now that tests and coverage are passing, it is time to come back to this. This should be done before approval.

Reply:
Since this PR is already gigantic, one option could be merging like this but immediately open a new one that implements the derivatives / variational circuits part. What do you think @BrunoLiegiBastonLiegi ?